packages/logging/src/metadata.js

```diff
@@ -0,0 +1,297 @@
+/*!
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module logging/metadata
+ */
+
+'use strict';
+var is = require('is');
+
+/**
+ * The abstract events emitted to invokers of getDefaultResource and
+ * getProjectId which are fired upon evaluation of the response from the
+ * metadata service.
+ * @const REQUEST_EVENTS
+ */
+var REQUEST_EVENTS = {
+  SUCCESS: 'METADATA_REQUEST_SUCCESS',
+  FAILURE: 'METADATA_REQUEST_FAILURE'
+};
+
+/**
+ * The Metadata class attempts to contact the metadata service and determine,
+ * based on request success and environment variables, what type of resource
+ * the library is operating on.
+ *
+ * @constructor
+ * @alias module:metadata
+ *
+ * @resource [Logs Resource API Documentation]{@link https://cloud.google.com/logging/docs/api/reference/rest/v2/MonitoredResource}
+ *
+ * @param {Duplex} requestStream - A stream returned by a request style
+ *  invocation to the metadata project id endpoint
+ */
+function Metadata(requestStream) {
+  this.metadataProjectId_ = null;
+  this.metadataContactError_ = null;
+  this.metadataRequestStream_ = requestStream;
+  this.attachRequestStreamListeners_(requestStream);
+}
+
+/**
+ * Attaches core request lifecycle event listeners on the request stream
+ * provided to the constuctor.
+ *
+ * @private
+ */
+Metadata.prototype.attachRequestStreamListeners_ = function(stream) {
+  var self = this;
```

Comment:
So, sorry. But I think making a callback style interface would cause more complexity. Given the case where we have multiple callers I still have to have a way of queuing callbacks on the resolution of the request to the metadata service and with the the callback-style request I would either have to create another event emitter (like the metadata class itself) to emit events too or create another single event to callback invoker on the class which is separate from the request itself. This way the request stream is the event emitter which unifies management of the event chain and reduces unneeded redundancy. This is also the same basic request functionality that the library itself exhibits.